IBD using chainwork instead of height and not using header timestamps

gmaxwell opened the pull request:
This introduces a 'minimum chain work' chainparam which is intended
 to be the known amount of work in the chain for the network at the
 time of software release.  If you don't have this much work, you're
 not yet caught up.
    
This is used instead of the count of blocks test from checkpoints.

This criteria is trivial to keep updated as there is no element of
 subjectivity, trust, or position dependence to it. It is also a more
 reliable metric of sync status than a block count.

This is the first step in a larger change to eliminate checkpoints
 completely, but it's independently a good idea.

This PR makes IsInitialBlockDownload no longer use header-only timestamps.

This avoids a corner case (mostly visible on testnet) where bogus
 headers can keep nodes in IsInitialBlockDownload.

gmaxwell committed:
IBD check uses minimumchain work instead of checkpoints.

This introduces a 'minimum chain work' chainparam which is intended
 to be the known amount of work in the chain for the network at the
 time of software release.  If you don't have this much work, you're
 not yet caught up.

This is used instead of the count of blocks test from checkpoints.

This criteria is trivial to keep updated as there is no element of
subjectivity, trust, or position dependence to it. It is also a more
reliable metric of sync status than a block count.

gmaxwell committed:
IsInitialBlockDownload no longer uses header-only timestamps.

This avoids a corner case (mostly visible on testnet) where bogus
 headers can keep nodes in IsInitialBlockDownload.

MarcoFalke commented:
utACK f3f8d2e

Nit: `GetTotalBlocksEstimate` is no longer used (only in tests). Feel free to remove that now.

Leviathn commented:
utACK - certainly a more reliable metric.

gmaxwell committed:
Remove GetTotalBlocksEstimate and checkpoint tests that test nothing.

GetTotalBlocksEstimate is no longer used and it was the only thing
 the checkpoint tests were testing.

Since checkpoints are on their way out it makes more sense to remove
 the test file than to cook up a new pointless test.

gmaxwell commented:
@MarcoFalke nits addressed, I think.

rebroad commented on src/chainparams.cpp:
```diff
@@ -96,6 +96,9 @@ class CMainParams : public CChainParams {
         consensus.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout = 1510704000; // November 15th, 2017.
 
+        // The best chain should have at least this much work.
+        consensus.nMinimumChainWork = uint256S("0x0000000000000000000000000000000000000000002cb971dd56d1c583c20f90");
```
Where is this number from please?

MarcoFalke replied:
@rebroad It is mentioned in the doc change in this very pull...

dcousens replied:
@MarcoFalke it mentions the source,  but not the conditions with which to re-create it.
I suppose I would just iterate the blockchain until I found it?

gmaxwell replied:
You can just run the getblockchaininfo rpc per the instructions and observe that its a number equal to or greater than that in order to verify it. :)

rebroad replied:
@MarcoFalke in the "doc change"? Not sure what you mean.

MarcoFalke commented:
re-utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

laanwj commented:
utACK e141beb

sipa commented:
utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

sipa commented:
I wonder if this should be backported to 0.13.2, as it fixes a visible issue on testnet.

jtimon commented:
utACK e141beb6a9816b7e1e680fb0a8bae16d42a3e557

mruddy commented on src/main.cpp:
```diff
@@ -1740,13 +1740,14 @@ bool IsInitialBlockDownload()
-    return state;
+    if (chainActive.Tip()->nChainWork < UintToArith256(chainParams.GetConsensus().nMinimumChainWork))
+        return true;
+    if (chainActive.Tip()->GetBlockTime() < (GetTime() - nMaxTipAge))
```
Any particular reason to use GetTime here over GetAdjustedTime?
